src/NLog/Config/LoggingConfiguration.cs

```diff
@@ -361,10 +369,12 @@ internal void Dump()
         internal void FlushAllTargets(AsyncContinuation asyncContinuation)
         {
             var uniqueTargets = new List<Target>();
-            foreach (var rule in this.LoggingRules)
+            for (int i = 0; i < this.LoggingRules.Count; i++)
```

Comment:
You have a good point here. We don't get the "System.InvalidOperationExceptionCollection was modified; enumeration operation may not execute.", but it's still not thread-safe